### PR TITLE
MPO

### PR DESCRIPTION
**Pull request by qgallouedec**
# What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/trl/tree/main/docs).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: trl/trainer/dpo_trainer.py [hunk 1]
```diff
@@ -1234,9 +1238,27 @@ def get_batch_loss_metrics(
...
+            loss_type = self.loss_type
+            loss_type_list = loss_type.split(",")
+
+            losses, chosen_rewards, rejected_rewards = 0, 0, 0
+            for curr_type in loss_type_list:
+                self.loss_type = curr_type
+                curr_losses, curr_chosen_rewards, curr_rejected_rewards = self.dpo_loss(
+                    model_output["chosen_logps"], model_output["rejected_logps"], ref_chosen_logps, ref_rejected_logps
+                )
+                curr_weight = getattr(self.args, f"{curr_type}_loss_weight")
```
Comment (ariG23498) [inline trl/trainer/dpo_trainer.py:1251]: This line of code assumes a loss weight in the `DPOConfig` which is unavailable.

Would it be okay to just add weighting parameter to the config?

**Comment by ariG23498** [comment]
Here is a rough [colab notebook](https://colab.research.google.com/drive/1uYm2r7kwcTDd6bUGGvdNpxNxzgtNSN1y?usp=sharing) that I have created.

In the notebook:
1. Installed trl from the branch
2. Used the `openbmb/RLAIF-V-Dataset` dataset
3. Used the `HuggingFaceTB/SmolVLM-Instruct` model

Here are my queries:
1. Reading the code, we expect the `loss_type` to be a str with `,` seperated losses.
    ```python
    training_args = DPOConfig(
        ...
        loss_type="sigmoid, bco_pair", # <-- a collection of losses
    )
    ```
    As far as I understand the MPO paper, it comprises of three losses, the DPO loss, the BCO loss, and the SFT loss. Here I have added the BCO and used sigmoid for the DPO loss, do we also have a way to add the SFT loss here somehow (which I think would be cross entropy loss?)
2. The weighting parameter in the config (as in this [comment](https://github.com/huggingface/trl/pull/2544#discussion_r1926549157))

File: trl/trainer/dpo_trainer.py [hunk 2]
```diff
@@ -1234,9 +1238,27 @@ def get_batch_loss_metrics(
         else:
             ref_chosen_logps, ref_rejected_logps = self.compute_ref_log_probs(batch)
 
-        losses, chosen_rewards, rejected_rewards = self.dpo_loss(
-            model_output["chosen_logps"], model_output["rejected_logps"], ref_chosen_logps, ref_rejected_logps
-        )
+        if "," in self.loss_type:
```
Comment (qgallouedec) [inline trl/trainer/dpo_trainer.py:1241]: Idelament, at this point, we would have
- `self.loss_type` a list of strings (eg, `[“sigmoid”, “bco_pair”]`)
- `self.loss_type_to_weights` a dict of `str, float` which for each loss type associates a weight.

Parsing for loss type could be done directly in the config, as here:

https://github.com/huggingface/trl/blob/fe4b5efe4e23f4331ba9c5b0c8bd92dc8302c287/trl/trainer/nash_md_config.py#L34-L46

File: trl/trainer/dpo_trainer.py (hunk 1 above)
Comment (qgallouedec) [reply trl/trainer/dpo_trainer.py:1251]: Adding a dict that associates each loss type to its weight would probably be better.

**Comment by ariG23498** [comment]
@qgallouedec Here is a rough [colab notebook](https://colab.research.google.com/drive/1uYm2r7kwcTDd6bUGGvdNpxNxzgtNSN1y?usp=sharing) with the current MPO training.

Let me know what you think.

File: trl/trainer/dpo_config.py [hunk 3]
```diff
@@ -282,7 +284,10 @@ class DPOConfig(TrainingArguments):
             "`transformers.TrainingArguments`."
         },
     )
-    loss_type: str = field(
+    loss_weights: Optional[Dict[str, float]] = field(
+        default_factory=lambda: ["your_values"]
+    )
+    loss_type: List[str] | str = field(
```
Comment (qgallouedec) [inline trl/trainer/dpo_config.py:290]: I need to check if this work with the parser and the cli

File: trl/trainer/dpo_trainer.py [hunk 4]
```diff
@@ -1294,9 +1302,36 @@ def get_batch_loss_metrics(
...
+            # Hold the list of losses
+            loss_type_list = self.loss_type
+
+            # Initialize the losses and rewards
+            losses, chosen_rewards, rejected_rewards = 0, 0, 0
+
+            # Iterate over the list of losses
+            for curr_loss_type in loss_type_list:
+                # Set the `loss_type` as the current loss, used for dpo_loss computation
+                self.loss_type = curr_loss_type
```
Comment (qgallouedec) [inline trl/trainer/dpo_trainer.py:1315]: Add an arg to dpo_loss instead

**Review by qgallouedec — COMMENTED** [review]
Looks good! Some initial comments :)

File: trl/trainer/dpo_config.py (hunk 3 above)
Comment (ariG23498) [reply trl/trainer/dpo_config.py:290]: If you could let me know how you might check it, I could do it and report back to you.

Comment (qgallouedec) [reply trl/trainer/dpo_config.py:290]: running this should work:

```
trl dpo --output_dir {tmp_dir} --model_name_or_path trl-internal-testing/tiny-Qwen2ForCausalLM-2.5 --dataset_name trl-internal-testing/zen --dataset_config standard_preference --report_to none
```

**Comment by qgallouedec** [comment]
Really nice!!!

But don't do this:

```python
    # Apply the chat template
    prompt = processor.apply_chat_template(prompt, tokenize=False)
    chosen = processor.apply_chat_template(chosen, tokenize=False)
    rejected = processor.apply_chat_template(rejected, tokenize=False)
```
The DPO Trainer handle applying the chat template. See #1930 for more info.

This code snippet is present in so many examples online, it's a scourge. 😩

File: trl/trainer/dpo_config.py (hunk 3 above)
Comment (ariG23498) [reply trl/trainer/dpo_config.py:290]: I tried the above command, which resulted in an issue with the `self.loss_type` being `None`.

Upon changing the command to `trl dpo --output_dir tmp_dir --model_name_or_path trl-internal-testing/tiny-Qwen2ForCausalLM-2.5 --dataset_name trl-internal-testing/zen --dataset_config standard_preference --report_to none --loss-type sigmoid` it started to train. I am not sure why this happens 🤔

File: trl/trainer/dpo_config.py [hunk 5]
```diff
@@ -103,7 +103,9 @@ class DPOConfig(TrainingArguments):
         learning_rate (`float`, *optional*, defaults to `1e-6`):
             Initial learning rate for [`AdamW`] optimizer. The default value replaces that of
             [`~transformers.TrainingArguments`].
-        loss_type (`str`, *optional*, defaults to `"sigmoid"`):
+        loss_weights (`dict`, *optional*, defaults to `{"sigmoid": 1.0}`):
+            Use to weight a combination of losses.
```
Comment (qgallouedec) [inline trl/trainer/dpo_config.py:107]: Maybe add that 
- the keys in `loss_weights` must be in `loss_type`
- by default (if not specified in the dict), the weight for a loss in `loss_type` is `1.0`

File: trl/trainer/dpo_config.py [hunk 6]
```diff
@@ -103,7 +103,9 @@ class DPOConfig(TrainingArguments):
         learning_rate (`float`, *optional*, defaults to `1e-6`):
             Initial learning rate for [`AdamW`] optimizer. The default value replaces that of
             [`~transformers.TrainingArguments`].
-        loss_type (`str`, *optional*, defaults to `"sigmoid"`):
+        loss_weights (`dict`, *optional*, defaults to `{"sigmoid": 1.0}`):
```
Comment (qgallouedec) [inline trl/trainer/dpo_config.py:106]:
```suggestion
        loss_weights (`dict`, *optional*, defaults to `None`):
```

File: trl/trainer/dpo_config.py [hunk 7]
```diff
@@ -282,7 +284,10 @@ class DPOConfig(TrainingArguments):
             "`transformers.TrainingArguments`."
         },
     )
-    loss_type: str = field(
+    loss_weights: Optional[Dict[str, float]] = field(
+        default_factory=lambda: ["your_values"]
+    )
```
Comment (qgallouedec) [inline trl/trainer/dpo_config.py:289]:
```suggestion
    loss_weights: Optional[dict[str, float]] = field(
        default=None,
        metadata={
            "help": "<exact same docstring as above>",
    )
```

File: trl/trainer/dpo_trainer.py [hunk 8]
```diff
@@ -428,6 +428,7 @@ def make_inputs_require_grad(module, input, output):
 
         self.beta = args.beta
         self.label_smoothing = args.label_smoothing
+        self.loss_weights = args.loss_weights
         self.loss_type = args.loss_type
```
Comment (qgallouedec) [inline trl/trainer/dpo_trainer.py:432]:
```suggestion
        self.loss_type = [args.loss_type] if not isinstance(args.loss_type, list) else args.loss_type
```

File: trl/trainer/dpo_trainer.py [hunk 9]
```diff
@@ -428,6 +428,7 @@ def make_inputs_require_grad(module, input, output):
 
         self.beta = args.beta
         self.label_smoothing = args.label_smoothing
+        self.loss_weights = args.loss_weights
```
Comment (qgallouedec) [inline trl/trainer/dpo_trainer.py:431]:
```suggestion
        self.loss_weights = args.loss_weights or {}
```

File: trl/trainer/dpo_trainer.py [hunk 10]
```diff
@@ -1294,9 +1302,36 @@ def get_batch_loss_metrics(
         else:
             ref_chosen_logps, ref_rejected_logps = self.compute_ref_log_probs(batch)
 
-        losses, chosen_rewards, rejected_rewards = self.dpo_loss(
-            model_output["chosen_logps"], model_output["rejected_logps"], ref_chosen_logps, ref_rejected_logps
-        )
+        if isinstance(self.loss_type, list):
```
Comment (qgallouedec) [inline trl/trainer/dpo_trainer.py:1305]: With the above suggestion, it is always a list

File: trl/trainer/dpo_trainer.py (hunk 4 above)
Comment (qgallouedec) [reply trl/trainer/dpo_trainer.py:1315]: yes

File: trl/trainer/dpo_config.py [hunk 11]
```diff
@@ -15,7 +15,7 @@
 import warnings
 from dataclasses import dataclass, field
 from enum import Enum
-from typing import Any, Callable, Optional, Union
+from typing import Any, Callable, Optional, Union, List, Dict
```
Comment (qgallouedec) [inline trl/trainer/dpo_config.py:18]: you can use `list` and `dict` instead

File: trl/trainer/dpo_config.py [hunk 12]
```diff
@@ -103,7 +103,10 @@ class DPOConfig(TrainingArguments):
         learning_rate (`float`, *optional*, defaults to `1e-6`):
             Initial learning rate for [`AdamW`] optimizer. The default value replaces that of
             [`~transformers.TrainingArguments`].
-        loss_type (`str`, *optional*, defaults to `"sigmoid"`):
+        loss_weights (`dict[str, float]` or `None`, *optional*, defaults to `None`):
+            Use to weight a combination of losses. The keys must be in `loss_type`. By default (if not specified in the dict),
+            the weight for a loss in loss_type is 1.0.
+        loss_type (`str` or `list`, *optional*, defaults to `"sigmoid"`):
```
Comment (qgallouedec) [inline trl/trainer/dpo_config.py:109]: can you also document that, when a list is passed, the loss is the sum of these values?

Comment (qgallouedec) [reply trl/trainer/dpo_config.py:109]: add `"sft"` as well

File: trl/trainer/dpo_trainer.py [hunk 13]
```diff
@@ -1294,9 +1303,22 @@ def get_batch_loss_metrics(
...
+        # Initialize the losses and rewards
+        losses, chosen_rewards, rejected_rewards = 0, 0, 0
+
+        # Iterate over the list of losses
+        for curr_loss_type in self.loss_type:
+            # Compute loss with current loss type
+            curr_losses, curr_chosen_rewards, curr_rejected_rewards = self.dpo_loss(
+                curr_loss_type, model_output["chosen_logps"], model_output["rejected_logps"], ref_chosen_logps, ref_rejected_logps
+            )
+            curr_loss_weight = getattr(self.loss_weights, curr_loss_type, 1.0)
```
Comment (qgallouedec) [inline trl/trainer/dpo_trainer.py:1315]: nice!

**Comment by ariG23498** [comment]
@qgallouedec I have resolved the merge conflicts and have also worked on the review suggestions. Could you help me with another round of review?

If this looks good, I can start a small training run on VLM with MPO. WDYT?

**Comment by ariG23498** [comment]
@qgallouedec a gentle ping here!

**Comment by cx9208** [comment]
can this merged? its helpful

File: trl/trainer/dpo_trainer.py [hunk 14]
```diff
@@ -1627,6 +1633,72 @@ def concatenated_forward(
...
+        """
+        Compute combined loss for multiple loss types (e.g., "sigmoid,bco_pair,sft" for MPO).
+
+        This enables combining different loss formulations with individual weights.
+        For example, MPO uses DPO (sigmoid) as preference loss, BCO (bco_pair) as quality loss,
+        and SFT (sft) as generation loss.
+        """
+        # Store original loss type
+        original_loss_type = self.loss_type
+        loss_type_list = original_loss_type.split(",")
```
Comment (ariG23498) [inline trl/trainer/dpo_trainer.py]: I think we should also add
```
loss_type_list = list(map(lambda x: x.strip(), loss_type_list))
```

The rationale behind it is if a user adds spaces between comma-separated loss functions, the script will not error out.

**Review by ariG23498 — COMMENTED** [review]
The implementation looks good to me.

File: trl/trainer/dpo_trainer.py (hunk 14 above)
Comment (sergiopaniego) [reply trl/trainer/dpo_trainer.py]: agree!

Comment (kashif) [inline trl/trainer/dpo_trainer.py]:
```suggestion
        loss_type_list = [loss_type.strip() for loss_type in original_loss_type.split(",")]
```

File: trl/trainer/dpo_trainer.py [hunk 15]
```diff
@@ -1627,6 +1633,72 @@ def concatenated_forward(
...
+        loss_type_list = [loss_type.strip() for loss_type in original_loss_type.split(",")]
+
+        # Initialize combined losses
+        combined_losses = 0
+        combined_chosen_rewards = 0
+        combined_rejected_rewards = 0
+
+        # Compute each loss type
+        for loss_type in loss_type_list:
+            self.loss_type = loss_type.strip()
```
Comment (kashif) [inline trl/trainer/dpo_trainer.py]:
```suggestion
            self.loss_type = loss_type
```
